LocalLang/Semantics.lean

```diff
@@ -1,12 +1,80 @@
 import LocalLang.AST
+import LocalLang.Evaluator
+
+inductive Ctx : Type where
+  | hole : Ctx
+  | bin_op_lhs : Ctx → BinOp → Expr → Ctx
+  | bin_op_rhs : ℕ → BinOp → Ctx → Ctx
+  | letin : String → Ctx → Expr → Ctx
+
+def Ctx.fill (e : Expr) : Ctx → Expr
+  | hole => e
+  | bin_op_lhs ctx op e' => Expr.binOp op (ctx.fill e) e'
+  | bin_op_rhs n op ctx => Expr.binOp op (.const n) (ctx.fill e)
+  | letin x ctx e' => .letIn x (ctx.fill e) e'
+
+abbrev Env := Std.HashMap String ℕ
+
+-- .funCall f
+def letin_chain (vars : List (String × Expr)) (e : Expr) : Expr := match vars with
+  | .nil => e
+  | .cons (name, expr) vs => .letIn name expr <| letin_chain vs e
```

Comment:
Instead of doing recursion directly, perhaps this could be a fold?  That way you can use library theorems about fold.